### DISCUSSION TITLE
Picard iteration divergence problem during MultiApps coupling

### DISCUSSION THREAD

**Discussion by WayneX-Wong** [idaholab/moose#23102] [Q&A General] [answered]
Hi all,

I was running a MultiApps problem,  main app is a heat conduction problem, and sub app is a simple lumped parameter problem. The calculation process is that main app uses the initial temperature boundary conditions to calculate steady-state heat conduction, after that, the integrate heat flux is postprocessed at the boundary and then transfer to the sub app to calculate and update the boundary temperature and continue the iteration.

I set Picard iteration in the [Executioner] block, use the default fixed_point_rel_tol, and set fixed_point_min_its to 2. The calculation converge in the second picard iteration, and the results look reasonable; However, when I set fixed_point_min_its to larger (10), it still converges after the tenth picard iteration, but the result is different and extremely unreasonable.

So I have two questions: 
1. Seems Picard iteration is not converged from the calculation results, then why it showed solve converged?
2. What parameters can be adjusted to improve the convergence of iteration? I've tried to set relaxation_factor to <1, and also refine the mesh or reduce the tolerance to enhanced convergence, but still no improvement.

Any guidance is greatly appreciated!

**Comment by GiudGiud** [comment, marked as answer]
Hello

Can you please attach the logs for these two runs?
One thing to note is that the criterion for convergence of these multiapps coupling is the residual in the main app. So if the subapp does not influence the main app much (can happen also if you are missing a transfer from the subapp to the main app), you can have a converged fixed point iteration between the apps, yet the subapp result is not converged. 

Relaxation factor is the only lever pretty much here. Everything else only deals with how converged the result should be.
Other fixed point iteration algorithms are more brittle than Picard as well

Guillaume

**Reply by WayneX-Wong** [reply]
Thanks for your reply Guillaume.
I attached the log files and main app input here.
[log_2_iterations.txt](https://github.com/idaholab/moose/files/10368412/log_2_iterations.txt)
[log_10_iterations.txt](https://github.com/idaholab/moose/files/10368414/log_10_iterations.txt)
[input.txt](https://github.com/idaholab/moose/files/10368415/input.txt)

I originally thought fixed point iteration is that the two fields exchange data in one Picard iteration step and converge relative to the previous step, so I thought that the convergence criterion of fixed point iteration was the field value difference between two iteration steps. Did I misunderstand?

**Reply by GiudGiud** [reply]
I should double check, but it should not be that. 
We have a very valid way of checking if something is converged, by computing the residual. We dont need to rely on the difference between two iterations. 

Two things:
```
 0 Picard |R| = 5.290898e+04
 1 Picard |R| = 2.343748e-07
 2 Picard |R| = 5.052297e-12
 3 Picard |R| = 4.752913e-04
 4 Picard |R| = 2.395892e-11
 5 Picard |R| = 3.797361e-07
 6 Picard |R| = 2.392503e-11
 7 Picard |R| = 2.137410e-09
 8 Picard |R| = 2.585106e-11
 9 Picard |R| = 4.846381e-09
10 Picard |R| = 3.554305e-11
```
the main app residual is jumping around between odd and even fixed point iterations/.

Similarly:
fixed point 5
```
HP80: Time Step 1, time = 1, dt = 1
HP80:  0 Nonlinear |R| = 1.128290e-06
HP80:  1 Nonlinear |R| = 2.371580e-10
HP80:  Solve Converged!
```

fixed point 6
```
HP80: Time Step 1, time = 1, dt = 1
HP80:  0 Nonlinear |R| = 1.832242e-01
HP80:  1 Nonlinear |R| = 4.521310e-05
HP80:  2 Nonlinear |R| = 1.649082e-08
HP80:  3 Nonlinear |R| = 3.225161e-12
HP80:  Solve Converged!
```

the residual of the subapps are also jumping around between fixed point iterations. in fact it seems to grow every other iteration.


This is symptomatic of either instability or a problem with the timing of execution of the transfers pipeline. 
Right now I see this:
```
  [boundary_temp_post2]
    type = ScalarVariable
    variable = boundary_temp2
    execute_on = TIMESTEP_BEGIN
  []
```
this will lag the postprocessor by one iteration. It might be the issue. Execute this more often by using the default or adding more execute_on

**Reply by hugary1995** [reply]
As you are probably already aware of, Picard isn't necessarily convergent: https://arxiv.org/pdf/1410.1390.pdf

The simplest way to tell if Picard is convergent is by looking at each sub-problem and check 
1. if the sequence of Picard iterates are bounded
2. if the sub-problem is convex

If both are true, then Picard should converge, though the convergence interval may vary.

**Reply by WayneX-Wong** [reply]
@GiudGiud I Change execute_on as default, some changes occurred, but Picard still didn't converge, and eventually main app didn't converge in a non-linear iteration step, and I break the running, as shown in the run log:
[log_default_trans_post.txt](https://github.com/idaholab/moose/files/10374147/log_default_trans_post.txt)

I tried adding more execute_on also, still not convergent.

**Reply by WayneX-Wong** [reply]
Thanks for your guidance @hugary1995 , I actually suspect that my coupling problem itself is divergent... I will read the paper of this link and try to find more clues.

**Reply by GiudGiud** [reply]
@WayneX-Wong one thing to try is to run a relaxation transient, with 'fake' time steps where only 1 fixed point iteration is taken on every time step.

Another thing is underrelaxation. 
I think you may need to specify which variables to relax (with relaxed_variables or transformed_variables parameters), it wont do all by default

**Reply by WayneX-Wong** [reply]
@GiudGiud I tried these things, and I've set a very small relaxation factor, but still no improvement.

I don't know yet how to mathematically analyze the convergence of this coupling problem, while I briefly analyzed from the physical aspect. 

The coupling parameters are the heat flux and boundary temperature, because the thermal conductivity of the problem is large and the geometry is small, small temperature differences can lead to large heat flux differences, which in turn make larger temperature differences, results to divergence.

So I changed the thermal conductivity directly to 1% of what it was before, and picard converge well without underrelaxation.

So I can probably infer that the problem itself doesn't converge. Do you have any suggestions?

**Reply by GiudGiud** [reply]
But did you specify which variables to relax?

Is the system itself physically unstable?

**Reply by WayneX-Wong** [reply]
Yes by declaring the transformed_variables parameter = temp. Can be seen that setting a smaller relaxation factor will cause the picard to have a tendency of convergence at the first few iterations, but still will diverge at specific iteration step, like
```
 0 Picard |R| = 1.234543e+05
 1 Picard |R| = 1.641611e-09
 2 Picard |R| = 1.539429e+04
 3 Picard |R| = 1.154572e+04
 4 Picard |R| = 8.735435e+03
 5 Picard |R| = 6.575765e+03
 6 Picard |R| = 4.931576e+03
 7 Picard |R| = 3.692373e+03
 8 Picard |R| = 2.752174e+03
 9 Picard |R| = 2.109329e+03
10 Picard |R| = 1.582342e+03
11 Picard |R| = 1.610063e+03
12 Picard |R| = 2.204031e+03
13 Picard |R| = 2.017019e+03
14 Picard |R| = 7.424021e+03
15 Picard |R| = 1.106489e+03
16 Picard |R| = 2.244228e+04
17 Picard |R| = 1.633168e+04
18 Picard |R| = 5.725446e+04
19 Picard |R| = 9.156200e+04
20 Picard |R| = 1.372499e+05
21 Picard |R| = 6.423847e+05
```

A larger relaxation factor leads to faster divergence, which implies to some extent, the relaxation factor is effective. But I guess this problem is physically unstable actually.

**Reply by GiudGiud** [reply]
You should try to use the conservative transfer parameters. This will reduce the impact of numerical errors during transfers

if the problem is physically unstable, you should try to add a constraint when solving it. Maybe force the integral of a quantity to be constant between each solve

**Reply by WayneX-Wong** [reply]
Sorry for taking so long to update this discussion.

I set the following parameters and the iterations eventually converge:

1. Set All **execute_on** to default (including which of [MultiApps] block)
2. Set **disable_fixed_point_residual_norm_check=true** , use **custom_pp** and relax the convergence criteria a little
3. Set **relaxation_factor** of main app (of [Executioner] block) to extremely small (around 0.1)

Seems that setting a smaller relaxation factor is indeed useful, maybe which was not small enough in my previous calculations.

Thanks again for your valuable advice! @GiudGiud @hugary1995
